SSL cache part 2

pluknet opened the pull request:
See commits.

pluknet committed:
SSL: caching certificates and certificate keys with variables.

A new directive "ssl_certificate_cache max=N [valid=time] [inactive=time]"
enables caching of SSL certificate chain and secret key objects specified
by "ssl_certificate" and "ssl_certificate_key" directives with variables.

Co-authored-by: Aleksei Bavshin <a.bavshin@nginx.com>

pluknet committed:
Upstream: caching certificates and certificate keys with variables.

Caching is enabled with proxy_ssl_certificate_cache and friends.

Co-authored-by: Aleksei Bavshin <a.bavshin@nginx.com>

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -175,12 +204,46 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+
+    old_cache = ngx_ssl_cache_get_old_conf(cf->cycle);
+
+    if (old_cache && old_cache->inherit) {
```
Shouldn't we check the current `cache->inherit` here?

pluknet replied:
There are two reasons why we should check rather a previous cycle configuration:
- logical
It is more consistent to state that configuration defines whether this configuration cycle should be prepared to inherit objects from *this* cycle to the next one.
Further, we can state that if the inheritance is not allowed from this cycle, we can optimize out such storage once parsing has finished (e.g., demolish rbtree, etc).
- technical
It is not possible to start using configuration while is in the middle of configuration parsing, because the merge phase did not yet happen.

bavshin-f5 replied:
I agree with the technical reason, but not fully convinced in logical:
* It means that the server state is not defined by the current config. I haven't encountered any directives in the open-source code explicitly affecting the next reload.
  There are some scenarios around `listen`, but that's more of a limitation that an intentional behavior.
* Two reloads are required to flush the cache.

At the very least, we should consider renaming `inherit` -> `inheritable` to point that it's not an instruction to the config parser but rather a sign of the old cache being inheritable.
Then we could try to draft the directive doc with various pitfalls and see if it can be explained clearly to the users.

jo-carter replied:
> technical
It is not possible to start using configuration while is in the middle of configuration parsing, because the merge phase did not yet happen.

Indeed, well not without requiring the user put the directive at the top of the configuration / prior to parsing of any other configuration that will call ngx_ssl_cache_fetch - load_module being an example of a directive that is obviously order dependent like this.

(_also because ngx_get_conf is used to access the configuration context, which works off cycle and not cf_)

>Two reloads are required to flush the cache.

This would be annoying.

pluknet replied:
Changed to inheritable.

pluknet replied:
> * It means that the server state is not defined by the current config. I haven't encountered any directives in the open-source code explicitly affecting the next reload.

This is unfortunate consequence of using the directive value when it cannot be used, that is, during the parsing.
Further, checking the value makes no sense after parsing has finished.

Anyway, this particular directive is a special one, because it alters state inheritance during configuration reload, so I believe existence of such directive with this specific behaviour is fine.

Another good option is to drop the directive.
There's no good reasons to keep it disabled, It was added to overcome potential environment or implementation issues.

pluknet replied:
> * Two reloads are required to flush the cache.

The directive isn't used to be altered frequently, so there is no issue. 

The value should either "on" (a reasonable default), or "off" to give a workaround in case something happens to not work as expected.

bavshin-f5 reviewed (commented):
> SSL: error message default in object caching API.

LGTM.

> SSL: encrypted certificate keys are exempt from object caching.

The change looks fine, but I have some concerns:
- I wonder if it is possible to restrict the effect to the keys that actually require passphrases. Is it even worth trying?
- Do we know if the configurations that prompted the SSL cache work use `ssl_password_file`? Would that regress the configuration load time for them?

> SSL: caching certificates and certificate keys with variables.
> Upstream: caching certificates and certificate keys with variables.

Not done looking through these, some preliminary comments:

I'm not sure if it's worth introducing both `valid` and `inactivity`  with a proactive expiration logic. If I understand the intention, open file cache has both to avoid wasting a limited resource &mdash; file descriptors. Here we can allow keeping a stale cert in memory to simplify the implementation.

pluknet commented:
> > SSL: error message default in object caching API.
> 
> LGTM.
> 

Actually, it doesn't belong to this series and should be fixed separately, preferably prior to the next release.
Currently this is the reason for potential segfaults, see bad malloc reports.
Moved to #333.

pluknet committed:
SSL: object cache inheritance from the old configuration cycle.

Memory based objects are always inherited, engine based objects are
never inherited to adhere the volatile nature of engines, file based
objects are inherited subject to modification time and file index.

The previous behaviour to bypass cache from the old configuration cycle
is preserved with a new directive "ssl_object_cache_inheritable off;".

pluknet committed:
SSL: encrypted certificate keys are exempt from object cache.

SSL object cache, as previously introduced in 1.27.2, did not take
into account encrypted certificate keys that might be unexpectedly
fetched from the cache regardless of the matching passphrase.  To
avoid this, caching of encrypted certificate keys is now disabled
based on the passphrase callback invocation.

A notable exception is encrypted certificate keys configured without
ssl_password_file.  They are loaded once resulting in the passphrase
prompt on startup and reused in other contexts as applicable.

pluknet commented:
> I wonder if it is possible to restrict the effect to the keys that actually require passphrases. Is it even worth trying?

That's a good point.
I rewrote the patch to actually restrict only encrypted keys, based on callback invocation.
In the previous approach, it was possible to disable caching at all e.g. with `ssl_password_file` at the top level.

> Do we know if the configurations that prompted the SSL cache work use ssl_password_file? Would that regress the configuration load time for them?

I'm not aware of how encrypted keys are widely used.
Disabling caching of encrypted keys still makes a noticeable speedup in my basic tests, because certificates itself are cached.

We can think about it later, if such requested arise.

pluknet commented:
> I'm not sure if it's worth introducing both `valid` and `inactivity` with a proactive expiration logic.

I'd rather compare this to shared SSL sessions that have a similar expiration logic.
Here it is used to mitigate the problem of keeping cached items at the maximum configured size once fully populated.

Though, I don't see a strong effect from using it here, and it somewhat duplicates the "valid=" time, so it may have sense to rethink this parameter.

pluknet commented:
Changes:
- logic to prevent caching encrypted keys is rewritten
- cache nodes are now created *after* SSL object, because this is not needed for non-cacheable SSL objects (read: encrypted keys)
- the first two patches are swapped for more clear diff

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -180,18 +285,122 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+                    return NULL;
+                }
+
+                if (ngx_file_info(cn->id.data, &fi) != NGX_FILE_ERROR) {
```
Can we reuse the result of a previous `ngx_file_info` call here?

pluknet replied:
Updated, thanks.
It is now reused either for new or revalidated NGX_SSL_CACHE_PATH entries.

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -180,18 +285,122 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+    if (cn != NULL) {
+        ngx_queue_remove(&cn->queue);
+
+        if (now - cn->created > cache->valid) {
```
As discussed via other channels, this check should be aware of the type, similarly to the `inheritable` logic.
E.g. for `NGX_SSL_CACHE_DATA` we should always reuse, for `NGX_SSL_CACHE_ENGINE` always discard, for `PATH` revalidate.

Possible change:
```diff
diff --git a/src/event/ngx_event_openssl_cache.c b/src/event/ngx_event_openssl_cache.c
index a3c619d72..f916758fe 100644
--- a/src/event/ngx_event_openssl_cache.c
+++ b/src/event/ngx_event_openssl_cache.c
@@ -290,6 +290,7 @@ ngx_ssl_cache_connection_fetch(ngx_ssl_cache_t *cache, ngx_pool_t *pool,
     void                  *value;
     time_t                 now;
     uint32_t               hash;
+    ngx_int_t              rc;
     ngx_file_info_t        fi;
     ngx_ssl_cache_key_t    id;
     ngx_ssl_cache_type_t  *type;
@@ -317,8 +318,22 @@ ngx_ssl_cache_connection_fetch(ngx_ssl_cache_t *cache, ngx_pool_t *pool,
         ngx_queue_remove(&cn->queue);
 
         if (now - cn->created > cache->valid) {
+            switch (cn->id.type) {
 
-            if (ngx_file_info(cn->id.data, &fi) == NGX_FILE_ERROR
+            case NGX_SSL_CACHE_PATH:
+                rc = ngx_file_info(cn->id.data, &fi);
+                break;
+
+            case NGX_SSL_CACHE_DATA:
+                cn->created = now;
+                goto found;
+
+            case NGX_SSL_CACHE_ENGINE:
+                rc = NGX_FILE_ERROR;
+                break;
+            }
+
+            if (rc == NGX_FILE_ERROR
                 || ngx_file_uniq(&fi) != cn->uniq
                 || ngx_file_mtime(&fi) != cn->mtime)
             {
```

pluknet replied:
Thanks for reminder, addressed in a different manner.

pluknet commented:
Changes:
- partially reverted 7a26242de to allow caching encrypted keys with passphrase prompt, and related code simplifications.
- made "inactive=" actually gradually expire old nodes

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -180,18 +284,122 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+    cn->created = now;
+    cn->value = value;
+
+    if (ngx_file_info(cn->id.data, &fi) != NGX_FILE_ERROR) {
```
1. ngx_file_info can be skipped for non-file cache entries
2. We haven't copied the `id.data` yet

```suggestion
    if (id.type == NGX_SSL_CACHE_PATH
        && ngx_file_info(id.data, &fi) != NGX_FILE_ERROR)
    {
```

pluknet replied:
Fixed, thanks.

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -180,18 +284,122 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+
+                    return NULL;
+                }
+
```
This block would erroneously cache the key if you replace unencrypted object with an encrypted one.

Suggested change:
```diff
@@ -327,17 +342,19 @@ ngx_ssl_cache_connection_fetch(ngx_ssl_cache_t *cache, ngx_pool_t *pool,

                 type->free(cn->value);

-                cn->value = type->create(&id, err, &data);
-                if (cn->value == NULL) {
+                value = type->create(&id, err, &data);
+                if (value == NULL || data == NGX_SSL_CACHE_DISABLED) {
                     ngx_rbtree_delete(&cache->rbtree, &cn->node);

                     cache->current--;

                     ngx_free(cn);

-                    return NULL;
+                    return value;
                 }

+                cn->value = value;
+
                 if (ngx_file_info(cn->id.data, &fi) != NGX_FILE_ERROR) {
                     cn->mtime = ngx_file_mtime(&fi);
                     cn->uniq = ngx_file_uniq(&fi);
```

pluknet replied:
Indeed, thanks.

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -151,12 +204,63 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+
+    /* try to use a reference from the old cycle */
+
+#if (NGX_SUPPRESS_WARN)
```
I don't believe this supposed to be conditional.

If `ngx_file_info` returns an error and an old cache entry exist, `uniq == ocn->uniq && mtime == ocn->mtime` below will be comparing with uninitialized data on stack.
If, for whatever reason, we get `ngx_file_info` error _twice_ on a readable file, we'll be comparing one set of uninitialized values with another.

pluknet replied:
That's true.
I don't see a better solution than to exempt such files from cache, because we cannot revalidate them later,
either for statically or dynamically configured.
Anyway, Clang remains dummy enough to require stub initialization (moved it outside of NGX_SSL_CACHE_PATH).

pluknet commented:
Most notably, this change reuses ngx_file_info() and improves its error handling; improves cache policy for runtime objects.
To be re-based/merged into appropriate existing commits.

pluknet committed:
SSL: avoid using mismatched certificate/key cached pairs.

This can happen with certificates and certificate keys specified
with variables due to partial cache update in various scenarios:
- cache expiration with only one element of pair evicted
- on-disk update with non-cacheable encrypted keys
- non-atomic on-disk update

The fix is to retry with fresh data on X509_R_KEY_VALUES_MISMATCH.

bavshin-f5 commented:
First, a general note: I feel like the logic in the dynamic cache introduction is getting too complex for a single commit.
Would it make sense to split it in 2 parts: basic dynamic cache implementation with reload from disk every time `cn->valid` expires, and another commit with the file info checks?

***

I've read the latest changes, and I disagree with equating the ability to revalidate an already cached entry with the ability to cache.
My rationale here:
* The connection cache is already opt-in, with an expectation that the user is aware of the configuration. It's unlikely that someone accidentally enables runtime cache for `engine:` keys, but doesn't want that to happen. I'd say, the expectation is exactly the opposite.
* There's not a lot of risk in keeping an object cached for `valid` even if we are unable to confirm that it wasn't updated and extend the life time past that. An ability to tune `valid=` or disable the cache for current context and your `ngx_ssl_cache_purge` patch cover all the negative cases I could imagine.
* Disabling caching for one of the two related objects should increase the likeliness of getting `X509_R_KEY_VALUES_MISMATCH`. We already have to do that for encrypted keys, no reason to add more special cases :(

***

I wonder if you have any other potential uses for `ngx_ssl_cache_purge`. Otherwise, I'd like to suggest an alternative: passing a flag like `NGX_SSL_CACHE_FORCE_UPDATE` to `ngx_ssl_cache_connection_fetch`. This way we can avoid reallocation of the cache node.
If you think the function already has too many arguments, we can make it a mask on `index`. There`s no less than 30 unused bits, might as well use a few.

pluknet commented:
Merged fixup commits to use as a new baseline.

pluknet commented:
Rebase.

pluknet committed:
SSL: cache revalidation of file based dynamic certificates.

Revalidation is based on file modification time and uniq file index,
and happens after the cache object validity time is expired.

pluknet commented:
Addressed @bavshin-f5 comments.

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -268,7 +503,16 @@ ngx_ssl_cache_lookup(ngx_ssl_cache_t *cache, ngx_ssl_cache_type_t *type,
+
+            if (!cache->max || ngx_time() - cn->accessed <= cache->inactive) {
+                return cn;
+            }
```
The following change comes in mind, which is seemingly equivalent, but
* saves one branch per lookup while reading the configuration
* respects our decision to never expire the `NGX_SSL_CACHE_DATA`
* reuses the cache node

Except... is it even worth removing inactive object at this point? We'll recheck the file attributes and possibly load the same object a few lines below, so maybe reverting the `ngx_ssl_cache_lookup` change is enough.

```diff
diff --git a/src/event/ngx_event_openssl_cache.c b/src/event/ngx_event_openssl_cache.c
index b4f8ab84d..64594fe0c 100644
--- a/src/event/ngx_event_openssl_cache.c
+++ b/src/event/ngx_event_openssl_cache.c
@@ -323,6 +323,10 @@ ngx_ssl_cache_connection_fetch(ngx_ssl_cache_t *cache, ngx_pool_t *pool,
             goto found;
         }
 
+        if (now - cn->accessed > cache->inactive) {
+            invalidate = 1;
+        }
+
         if (!invalidate && now - cn->created <= cache->valid) {
             goto found;
         }
@@ -503,16 +507,7 @@ ngx_ssl_cache_lookup(ngx_ssl_cache_t *cache, ngx_ssl_cache_type_t *type,
         rc = ngx_memn2cmp(id->data, cn->id.data, id->len, cn->id.len);
 
         if (rc == 0) {
-
-            if (!cache->max || ngx_time() - cn->accessed <= cache->inactive) {
-                return cn;
-            }
-
-            ngx_ssl_cache_node_free(&cache->rbtree, cn);
-
-            cache->current--;
-
-            return NULL;
+            return cn;
         }
 
         node = (rc < 0) ? node->left : node->right;
```

pluknet replied:
I think we don't even need to remove/invalidate them here. This is an artifact after reusing SSL session cache model (where it actually makes sense).
Just reverting the `ngx_ssl_cache_lookup` change seems to be enough, it makes expired items to be limited to its intended purpose - to ngx_ssl_cache_expire() to give more room in cache.

bavshin-f5 commented on src/event/ngx_event_openssl_cache.c:
```diff
@@ -180,18 +283,150 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+
+        default:
+            ngx_log_debug1(NGX_LOG_DEBUG_CORE, pool->log, 0,
+                           "cached ssl file changed: %s", cn->id.data);
```
This isn't always file, so maybe "cached ssl object"? Same suggestion for `ngx_ssl_cache_node_free`

pluknet replied:
Agreed.
On a related note, we cannot always detect an actual change and may come here due to the limited lifetime.
So I'd rather change it to "update cached ssl ..", also to match "delete cached ssl .." in `ngx_ssl_cache_node_free`.

pluknet commented:
Addressed @bavshin-f5 comments.
- backed out change in ngx_ssl_cache_lookup
- adjusted comments

pluknet commented:
password callback fixes:
- MSVC C4701 bogus warning 
- initialization order restored

bavshin-f5 reviewed (approved):
Looks good!
